engine/src/main/java/org/terasology/registry/DynamicInstanceProvider.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright 2016 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.registry;
+
+import org.terasology.context.Context;
+import org.terasology.naming.Name;
+
+/**
+ * A DynamicInstanceProvider provides an instance dynamically.
+ */
+public interface DynamicInstanceProvider<T> {
```

Comment:
`DynamicInstanceProvider` is kind of a rendundant name as a `Provider` is already someting that provides you dyamically with instances. Also the javadoc currently is kind of redundant to the class name.

When you just look at this class as another developer you would do very hard to guess why we need it and what to do with it.

I think the idea behind it should be described at one of the classes like this one and other API parts should refer to this class for an explanation. Also the class name should give a better indication what the class is good for. For example the class could be called `ModuleSecurityBoundInstanceProvider`.

Example for javadoc:

> Some managers need to be able to check, if a using module has actually the permission to use a certain feature. The idea is to do this check at injection time. At this point of time the module of the class that gets injected can be determined. With this information it is possible to inject a module specific instance to the class. The creation or determination of this module specific instance gets done via this interface. Please note that implementors of this class must make sure that there is no way for other modules to obtain a instance of this interface and by pass thus the injection.
> 
> Of course modules that let a instance be injected should ensure that there is no way to access this instance from outside. This should be document in each type that can be obtained via this interface.